lib/Sema/ConstraintSystem.cpp

```diff
@@ -3709,6 +3709,141 @@ static bool diagnoseAmbiguity(
   return diagnosed;
 }
 
+using Fix = std::pair<const Solution *, const ConstraintFix *>;
+
+// Attempts to diagnose function call ambiguities of types inferred for a result
+// generic parameter from contextual type and a closure argument that
+// conflicting infer a different type for the same argument. Example:
+//   func callit<T>(_ f: () -> T) -> T {
+//     f()
+//   }
+//
+//   func context() -> Int {
+//     callit {
+//       print("hello")
+//     }
+//   }
+// Where generic argument `T` can be inferred both as `Int` from contextual
+// result and `Void` from the closure argument result.
+static bool
+diagnoseContextualFunctionCallGenericAmbiguity(ConstraintSystem &cs,
+                                               ArrayRef<Fix> contextualFixes,
+                                               ArrayRef<Fix> allFixes) {
+
+  if (contextualFixes.empty())
+    return false;
+
+  auto contextualFix = contextualFixes.front();
+  if (!std::all_of(contextualFixes.begin() + 1, contextualFixes.end(),
+                   [&contextualFix](Fix fix) {
+                     return fix.second->getLocator() ==
+                            contextualFix.second->getLocator();
+                   }))
+    return false;
+
+  auto fixLocator = contextualFix.second->getLocator();
+  auto contextualAnchor = fixLocator->getAnchor();
+  auto *AE = getAsExpr<ApplyExpr>(contextualAnchor);
+  // All contextual failures anchored on the same function call.
+  if (!AE)
+    return false;
+
+  auto fnLocator = cs.getConstraintLocator(AE->getSemanticFn());
+  auto overload = contextualFix.first->getOverloadChoiceIfAvailable(fnLocator);
+  if (!overload)
+    return false;
+
+  auto applyFnType = overload->openedType->castTo<FunctionType>();
+  auto resultTypeVar = applyFnType->getResult()->getAs<TypeVariableType>();
+  if (!resultTypeVar)
+    return false;
+
+  auto *GP = resultTypeVar->getImpl().getGenericParameter();
+  if (!GP)
+    return false;
+
+  auto typeParamResultInvolvesTypeVar =
+      [&applyFnType](unsigned paramIdx, TypeVariableType *typeVar) {
+        auto param = applyFnType->getParams()[paramIdx];
+        auto paramType = param.getParameterType()->castTo<FunctionType>();
+
+        bool contains = false;
+        paramType->getResult().visit([&](Type ty) {
```

Comment:
There is a helper for this already - `ConstraintSystem::typeVarOccursInType` since it's a common operation during inference...